### PR TITLE
Fix the problem of int overflow when trying to generate inverted index on huge number of entries

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
Added checks before generating inverted index.
If the check failed, log a warning and skip creating the inverted index.

### CODE REVIEW

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java [hunk 1]
```diff
@@ -167,14 +167,38 @@ public void init(SegmentGeneratorConfig segmentCreationSpec, SegmentIndexCreatio
     }
 
     for (String column : config.getInvertedIndexCreationColumns()) {
-      if (!schema.hasColumn(column)) {
-        LOGGER.warn("Skipping enabling index on column:{} since its missing in schema", column);
+      // Validation check
+      FieldSpec fieldSpec = schema.getFieldSpecFor(column);
+      if (fieldSpec == null) {
+        LOGGER.warn("Skip creating inverted index for segment: {}, column: {} because column does not exist in schema",
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java]: is it possible to log the segment or schema name?

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java [hunk 2]
```diff
@@ -167,14 +167,38 @@ public void init(SegmentGeneratorConfig segmentCreationSpec, SegmentIndexCreatio
...
+            segmentName, column);
+        continue;
+      }
+      if (totalDocs > OffHeapBitmapInvertedIndexCreator.MAX_NUM_ENTRIES) {
+        LOGGER.warn(
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java]: pls log the segment or schema name

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java [hunk 3]
```diff
@@ -167,14 +167,38 @@ public void init(SegmentGeneratorConfig segmentCreationSpec, SegmentIndexCreatio
...
         continue;
       }
       ColumnIndexCreationInfo indexCreationInfo = indexCreationInfoMap.get(column);
-      int uniqueValueCount = indexCreationInfo.getDistinctValueCount();
-      OffHeapBitmapInvertedIndexCreator invertedIndexCreator = new OffHeapBitmapInvertedIndexCreator(file,
-          uniqueValueCount, totalDocs, indexCreationInfo.getTotalNumberOfEntries(), schema.getFieldSpecFor(column));
+      int cardinality = indexCreationInfo.getDistinctValueCount();
+      if (cardinality > OffHeapBitmapInvertedIndexCreator.MAX_NUM_ENTRIES) {
+        LOGGER.warn(
+            "Skip creating inverted index for segment: {}, column: {} because cardinality: {} exceeds the limit: {}",
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java]: pls log the segment or schema name

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java [hunk 4]
```diff
@@ -167,14 +167,38 @@ public void init(SegmentGeneratorConfig segmentCreationSpec, SegmentIndexCreatio
...
+            segmentName, column, cardinality, OffHeapBitmapInvertedIndexCreator.MAX_NUM_ENTRIES);
+        continue;
+      }
+      int totalNumberOfEntries = indexCreationInfo.getTotalNumberOfEntries();
+      if ((!fieldSpec.isSingleValueField())
+          && (totalNumberOfEntries > OffHeapBitmapInvertedIndexCreator.MAX_NUM_ENTRIES)) {
+        LOGGER.warn(
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java]: pls log the segment or schema name

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java (hunk 1 above)
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java]: We already logged the segment name.

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/loader/invertedindex/InvertedIndexHandler.java [hunk 5]
```diff
@@ -91,7 +91,31 @@ public void createInvertedIndices()
...
     String column = columnMetadata.getColumnName();
+    int totalDocs = columnMetadata.getTotalDocs();
+    if (totalDocs > OffHeapBitmapInvertedIndexCreator.MAX_NUM_ENTRIES) {
+      LOGGER.warn(
+          "Skip creating inverted index for segment: {}, column: {} because totalDocs: {} exceeds the limit: {}",
+          segmentName, column, totalDocs, OffHeapBitmapInvertedIndexCreator.MAX_NUM_ENTRIES);
+      return;
+    }
+    int cardinality = columnMetadata.getCardinality();
+    if (cardinality > OffHeapBitmapInvertedIndexCreator.MAX_NUM_ENTRIES) {
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/loader/invertedindex/InvertedIndexHandler.java]: Duplicated code logic. Move this to OffheapBitmapInvertedIndexCreator.isConstructionPossible(cardinality, totalDocs)  static method. That will keep all the logic in one place

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/inv/OffHeapBitmapInvertedIndexCreator.java [hunk 6]
```diff
@@ -71,6 +71,11 @@
   private static final Logger LOGGER =
       LoggerFactory.getLogger(OffHeapBitmapInvertedIndexCreator.class);
 
+  // Number of bytes required to store an integer
+  private static final int INT_SIZE = Integer.SIZE / Byte.SIZE;
+  // Maximum number of entries for each buffer
+  public static final int MAX_NUM_ENTRIES = Integer.MAX_VALUE / INT_SIZE;
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/inv/OffHeapBitmapInvertedIndexCreator.java]: This is not accounting the size correctly. Each entry takes Integer.size + bitmap serialization size. Bitmap serialization size is unpredictable but it's atleast 1 byte but most likely more. So, say, atleast 5 bytes per entry.
So, max entries can not exceed (Integer.max_value / (int_size + 1) - 1). For -1 see class comment

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/loader/invertedindex/InvertedIndexHandler.java (hunk 5 above)
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/loader/invertedindex/InvertedIndexHandler.java]: I did it this way for minor performance gain (which is definitely unnecessary :-P), will address this in a separate PR

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/inv/OffHeapBitmapInvertedIndexCreator.java (hunk 6 above)
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/inv/OffHeapBitmapInvertedIndexCreator.java]: The actual size calculation is trickier. For multi-value columns, we allow this number of entries; for number of total docs, we allow this number of total docs. But we don't allow this high cardinality, which is what actually stored in the final bitmap file.

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/loader/invertedindex/InvertedIndexHandler.java [hunk 7]
```diff
@@ -91,7 +91,31 @@ public void createInvertedIndices()
 
   private void createInvertedIndexForColumn(ColumnMetadata columnMetadata)
       throws IOException {
+    // Validation check
     String column = columnMetadata.getColumnName();
+    int totalDocs = columnMetadata.getTotalDocs();
+    if (totalDocs > OffHeapBitmapInvertedIndexCreator.MAX_NUM_ENTRIES) {
+      LOGGER.warn(
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/loader/invertedindex/InvertedIndexHandler.java]: I agree it's hard to track this performance problem if inverted indexes are not created. Actually adding a counter (metric) here does not help that much, and it's not very possible to pass the ServerMeter all the way down to this method. Should we just throw the exception, and let CRC check catch this?

**Pull request merged**